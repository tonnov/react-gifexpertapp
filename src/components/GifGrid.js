import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem'
import PropTypes from 'prop-types';
import 'animate.css'

export const GifGrid = ({category}) => {

    // const [images, setImages] = useState([]);

    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>
            
            { loading  && <p className='animate__animated animate__flash'>Cargando...</p> }

            <div className='card-grid'>
                {
                    images.map((item) => (
                        <GifGridItem 
                            key={item.id}
                            {...item}
                        />
                    ))
                }
            </div>
                    
            
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}