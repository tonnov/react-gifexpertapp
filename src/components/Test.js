import React, { useEffect, useState } from 'react'
import data from '../mock/data.json';

const getRemoteData = () => {
    // try {
    //     const response = await fetch('https://cdn.glitch.com/760886e0-5f1f-4216-9a2e-0e0c7c7797eb%2Fbuilds.json?v=1602261849911');
    //     const data = await response.json();
    //      return data;
    // } catch (error) {
    //     console.log(error)
    // }
    return data;
}

const getFails = (data) => {
    if(!data) return;
    return data.filter( item => item.result !== 'SUCCESS')
}

const getFailPct = (data) => {
    const fails = getFails(data);
    const total = data.length;
    const failsCount = fails.length;
    const pct = (failsCount / total) * 100;
    return pct;
}

export const Test = () => {
    const [ dataset, setDataset ] = useState(null);
    const [ failPct, setFailPct] = useState(0);
    const [leadTime, setLeadTime] = useState({});

    useEffect(() => {
        const runAsync = async () => {
            const data = await getRemoteData();
            setDataset(data);
        }

        runAsync();
    }, [])

    useEffect(() => {
        if (!dataset) return;

        const pct = getFailPct(dataset);
        setFailPct(pct);

        const initialState = {
            total: 0,
            duration: 0
        }

        const totalSucceed = dataset.reduce((acc, item) => {
            if (item.result === 'SUCCESS') {
                acc = {
                    total: acc.total + 1,
                    duration: acc.duration + item.duration
                }
            }
            return acc;
        }, initialState);

        const totalFailed = dataset.reduce((acc, item) => {
            if (item.result !== 'SUCCESS') {
                acc = {
                    total: acc.total + 1,
                    duration: acc.duration + item.duration
                }
            }
            return acc;
        }, initialState);

        const ms =  1000*Math.round((totalSucceed.duration)/1000);
        const d = new Date(ms);
        console.log( d.getUTCHours() + ':' + d.getUTCMinutes() + ':' + d.getUTCSeconds() );

        const success = Math.ceil(totalSucceed.duration / totalSucceed.total / 60000 );
        const fails = Math.ceil(totalFailed.duration / totalFailed.total / 60000 );

        console.log(totalSucceed);

        setLeadTime({ success, fails });

    },[dataset])

    // console.log(dataset);
    return (
        <>
            <h3>Change Fail Percentage</h3>
            <p>{failPct.toFixed(2)}%</p>
            <h3>Lead time for changes</h3>
            <ul>
                <li>Success Time: {leadTime.success} Minutes</li>
                <li>Fail Time: {leadTime.fails} Minutes</li>
            </ul>
        </>
    )
}

export default Test
