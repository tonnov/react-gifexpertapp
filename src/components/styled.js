import styled from 'styled-components'

export const InputBox = styled.input`
    width: 100%;
    height: 35px;
    outline: none;
    border: 2px solid #928fa7;
    border-radius: 4px;
    font-size: 12pt;
    font-weight: bold;
    padding: 2px 10px;
    color: #1555ea;

    ::placeholder {
        color: #ab9cb4;
    }
`;