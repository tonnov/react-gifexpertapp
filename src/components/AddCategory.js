import React, { useState } from 'react'
import PropTypes from 'prop-types';
import { InputBox } from './styled'

export const AddCategory = ({setCategories}) => {
    const [inputValue, setInputValue] = useState('')

    const handleSubmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {            
            setCategories(cats => [inputValue, ...cats]);
            setInputValue('');
        }
    }
    const handleInputChange = (e) => {
        setInputValue(e.target.value)
    }
    
    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <InputBox 
                type="text"
                value={inputValue}
                onChange={handleInputChange}
                placeholder="Add a new category..."
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}