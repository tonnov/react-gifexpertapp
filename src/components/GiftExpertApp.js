import React, { useState } from 'react'
import { AddCategory } from './AddCategory'
import { GifGrid } from './GifGrid'

export const GiftExpertApp = () => {

    const [categories, setCategories] = useState(['matrix']);

    // const handleAdd = () => {
    //     const rnd = Math.floor(Math.random() * 100);
    //     setCategories( cat => [...cat, `val ${rnd}`]);
    // }

    return (
        <>
            <h2>GiftExpertApp</h2>
            <AddCategory setCategories={setCategories} />
            {/* <hr /> */}
            <ol>
                { 
                    categories.map( category => (
                        <GifGrid key={category} category={category} />
                    ))
                }
            </ol>
        </>
    )
}
