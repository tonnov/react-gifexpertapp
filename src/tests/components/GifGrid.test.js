import React from 'react'
import { shallow } from 'enzyme'
import { GifGrid } from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs');

describe('Test for GifGrid', () => {

    const category = 'matrix';

    test('should render properly', () => {
        
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    })

    test('should test custom hook', () => {

        const gifs = [{
            id: 1,
            url: "https://imagenes.com/mygif.gif",
            title: "titulo",
        }]
        
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);

    })
    
    
    
})
