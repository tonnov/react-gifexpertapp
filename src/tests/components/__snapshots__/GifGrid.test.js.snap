// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for GifGrid should render properly 1`] = `
<Fragment>
  <h3>
    matrix
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test for GifGrid should test custom hook 1`] = `
<Fragment>
  <h3>
    matrix
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id={1}
      key="1"
      title="titulo"
      url="https://imagenes.com/mygif.gif"
    />
  </div>
</Fragment>
`;
